pkg/thumbnail/storage/filesystem.go

```diff
@@ -74,14 +70,68 @@ func (s FileSystem) BuildKey(r Request) string {
 	filetype := r.Types[0]
 	filename := r.Resolution.String() + "." + filetype
 
-	key := new(bytes.Buffer)
-	key.WriteString(etag[:2])
-	key.WriteRune('/')
-	key.WriteString(etag[2:4])
-	key.WriteRune('/')
-	key.WriteString(etag[4:])
-	key.WriteRune('/')
-	key.WriteString(filename)
+	return filepath.Join(etag[:2], etag[2:4], etag[4:], filename)
+}
+
+func (s FileSystem) rootDir(key string) string {
+	p := strings.Split(key, string(os.PathSeparator))
+	return p[0]
+}
+
+func (s FileSystem) storeImage(key string, img []byte) (string, error) {
```

Comment:
How about making this method thread safe by adding a `sync.Mutex` and acquiring on every write? If this method runs on a goroutine, they should be thread safe